crates/transport/src/pinnable.rs

```diff
@@ -0,0 +1,122 @@
+use std::pin::Pin;
+use std::{fmt, ops};
+
+/// A wrapper for [`parking_lot::Mutex`] that supports obtaining `Pin<&mut T>` references to the contained value.
+///
+/// [`parking_lot::Mutex<T>`] itself does not have structural pinning.
+/// The pinned-ness of the mutex type does not propagate to the field (the `T`).
+///
+/// `Pin<PinnableMutex<T>>` however can return a `Pin<MutexGuard<T>>`.
+/// It’s a trade-off, because it can no longer provide mutable access without being pinned.
+///
+/// # Example
+///
+/// ```
+/// use std::future::Future;
+/// use std::pin::Pin;
+/// use std::sync::Arc;
+/// use std::task::{Context, Poll};
+/// use transport::PinnableMutex;
+///
+/// fn poll_shared_future<F: Future>(
+///     fut: &Pin<Arc<PinnableMutex<F>>>,
+///     ctx: &mut Context<'_>,
+/// ) -> Poll<F::Output> {
+///     fut.as_ref().lock().as_mut().poll(ctx)
+/// }
+/// ```
+pub struct PinnableMutex<T: ?Sized>(parking_lot::Mutex<T>);
+
+impl<T: ?Sized> fmt::Debug for PinnableMutex<T>
+where
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&self.0, f)
```

Comment:
Question: Doesn't `#[derive(Debug)]` produce the same output?

Reply:
It's not derivable because T does not necessarily implement Debug. Unless they improved that recently?

Reply:
Huh,  it compiled successfully on playground at least https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=238783f816438c16430a790e0df7ddb4